pints/_log_pdfs.py

```diff
@@ -115,6 +115,161 @@ def sample(self, n=1):
+    (i.e. have the same value for each LogPDF), so that the resulting combined
+    LogPDF has :math:`m (n - k) + k` parameters.
+
+    This is useful for e.g. modelling the time-series of multiple individuals
+    (each individual defines a separate :class:`LogPDF`), and some parameters
+    are expected to be the same across individuals (for example, the noise
+    parameter across different individuals within the same experiment).
+
+    For two :class:`LogPDFs<pints.LogPDF>` :math:`L _1`  and
+    :math:`L _2` with parameters :math:`(\psi _1, \theta _1`) and
+    :math:`(\psi _2, \theta _2`) respectively, a pooling of the
+    :math:`\theta _i` results in a pooled log-pdfs of the form
+
+    .. math::
+        L(\psi _1, \psi _2, \theta | D_1, D_2) =
+            L(\psi _1, \theta | D_1) + L(\psi _2, \theta | D_2),
+
+    where :math:`\theta := \theta _1 = \theta _2`, and :math:`D_i` is the
+    measured time-series of individual :math:`i`.
+
+    The parameters in a :class:`PooledLogPDF` are ordered such that unpooled
+    parameters come before pooled parameters.
+
+    Extends :class:`LogPDF`.
+
+    Parameters
+    ----------
+    log_pdfs
+        A sequence of :class:`LogPDF` objects.
+    pooled
+        An array-like object of dtype bool, indicating which parameters across
+        the likelihoods are pooled (``True``) or remain unpooled (``False``).
+
+    Example
+    -------
+    ::
+
+        pooled_log_likelihood = pints.PooledLogPDF(
+            log_pdfs=[
+                pints.GaussianLogLikelihood(problem1),
+                pints.GaussianLogLikelihood(problem2)],
+            pooled=[False, True])
+    """
+    def __init__(self, log_pdfs, pooled):
+        super(PooledLogPDF, self).__init__()
+
+        # Check input arguments
+        if len(log_pdfs) < 2:
+            raise ValueError(
+                'PooledLogPDF requires at least two log-pdfs.')
+        for index, pdf in enumerate(log_pdfs):
+            if not isinstance(pdf, LogPDF):
+                raise ValueError(
+                    'All log-pdfs passed to PooledLogPDFs must be instances of'
+                    ' pints.LogPDF (failed on argument '
+                    + str(index) + ').')
+
+        # Check parameter dimension across log-pdfs
+        self._log_pdfs = log_pdfs
+        n_parameters = self._log_pdfs[0].n_parameters()
+        for pdf in self._log_pdfs:
+            if pdf.n_parameters() != n_parameters:
+                raise ValueError(
+                    'All log-pdfs passed to PooledLogPDFs must have '
+                    'same dimension.')
+
+        # Check that pooled matches number of parameters
+        self._pooled = np.asarray(pooled)
+        if len(self._pooled) != n_parameters:
+            raise ValueError(
+                'The array-like input `pooled` needs to have the same length '
+                'as the number of parameters of the individual log-pdfs.')
```

Comment:
Be good to check the Falses in `pooled` comes at the end? Since your docs ([here](https://github.com/pints-team/pints/pull/1212/files#diff-b121974bb553002da6fe16f041d9eaa6R142-R143)) say "The parameters in a :class:`PooledLogPDF` are ordered such that unpooled parameters come before pooled parameters." -- sounds important and easy to miss or implemented incorrectly. If it's not a strict requirement, then take that out.

Reply:
In particular we calculate the unpooled with `np.sum(~self._pooled)`, which does not care where the Falses are located!

Reply:
The pooled mask is wrt the parameters of the individual likelihoods, and you are free to pool any of the parameters, i.e. pooled = [True, False] is allowed. It means that the first parameter is pooled across all individuals. However for the input parameters of the pooled log-pdf the unpooled parameters of the individuals would come before the pooled parameters. Does that make sense?

Reply:
Hmm... I don't know, what kind of annoys me is that it looks like "this method is very general, but not that general"...

Reply:
I see, I must have misread the docstring as "LogPDF" instead of "PooledLogPDF", sorry!

Reply:
And probably because of the notation in the example "psi and theta" tricks me into thinking they must be ordered too...!

Reply:
This re-ordering of the parameters can be tricky! Because if I had `L_1(psi_{1,1}, theta_1, psi_{1,2})` and `L_2(psi_{2,1}, theta_1, psi_{2,2})`, this would give the final parameter as `psi_{1,1}, psi_{1,2}, psi_{2,1}, psi_{2,2}, theta_1`. This needs to be really really clear! (As in it is _not_ `psi_{1,1}, psi_{2,1}, psi_{1,2}, psi_{2,2}, theta_1` or anything else!) Because by the end, elsewhere in PINTS, we would only refer to it as parameter_1, 2, 3, 4, and 5, and we would have no idea how this is re-ordered.

Reply:
Agreed. So you'd say that 

"The parameters in a :class:`PooledLogPDF` are ordered such that unpooled
    parameters come before pooled parameters. The order of the unpooled
    parameters is determined by the order of the individual likelihoods."

is not explicit enough?

Reply:
I don't know to be honest, as I (the user) am no longer the one who explicitly define the order of the parameters but the class (as elsewhere I am the one defining all the order of model parameters!), and I have no way to check what these reordered parameters are afterwards in Pints (I don't think I can!). So this makes me worried. Maybe an example with a few more parameters (just like the one I showed above) would help...? @MichaelClerx what do you think?

Reply:
I agree the docstring needs to be very clear on this, with a clear description and an example (or two?)

Reply:
@chonlei I tried to make it as clear as possible. Let me know what you think

Reply:
@DavAug Thank you, with the example, it is much clearer now! I have made some slight tweaking to make it a bit more explicit. Make sure you git pull before git committing!

Reply:
I am happy now. Thanks!